src/FitterBase.cc

```diff
@@ -0,0 +1,221 @@
+#include <vector>
+
+#include "lsst/log/Log.h"
+
+#include "lsst/jointcal/Chi2.h"
+#include "lsst/jointcal/CcdImage.h"
+#include "lsst/jointcal/Eigenstuff.h"
+#include "lsst/jointcal/FitterBase.h"
+#include "lsst/jointcal/FittedStar.h"
+#include "lsst/jointcal/MeasuredStar.h"
+
+namespace {
+LOG_LOGGER _log = LOG_GET("jointcal.Fitter");
+}
+
+namespace lsst {
+namespace jointcal {
+
+Chi2Statistic FitterBase::computeChi2() const {
+    Chi2Statistic chi2;
+    accumulateStatImageList(_associations.getCcdImageList(), chi2);
+    accumulateStatRefStars(chi2);
+    // chi2.ndof contains the number of squares.
+    // So subtract the number of parameters.
+    chi2.ndof -= _nParTot;
+    return chi2;
+}
+
+unsigned FitterBase::findOutliers(double nSigmaCut, MeasuredStarList &msOutliers,
+                                  FittedStarList &fsOutliers) const {
+    // collect chi2 contributions
+    Chi2List chi2List;
+    chi2List.reserve(_nMeasuredStars + _associations.refStarList.size());
+    // contributions from measurement terms:
+    accumulateStatImageList(_associations.ccdImageList, chi2List);
+    // and from reference terms
+    accumulateStatRefStars(chi2List);
+
+    // compute some statistics
+    size_t nval = chi2List.size();
+    if (nval == 0) return 0;
+    sort(chi2List.begin(), chi2List.end());
+    double median = (nval & 1) ? chi2List[nval / 2].chi2
+                               : 0.5 * (chi2List[nval / 2 - 1].chi2 + chi2List[nval / 2].chi2);
+    auto averageAndSigma = chi2List.computeAverageAndSigma();
+    LOGLS_DEBUG(_log, "RemoveOutliers chi2 stat: mean/median/sigma " << averageAndSigma.first << '/' << median
+                                                                     << '/' << averageAndSigma.second);
+    double cut = averageAndSigma.first + nSigmaCut * averageAndSigma.second;
+    /* For each of the parameters, we will not remove more than 1
+       measurement that contributes to constraining it. Keep track using
+       of what we are touching using an integer vector. This is the
+       trick that Marc Betoule came up to for outlier removals in "star
+       flats" fits. */
+    Eigen::VectorXi affectedParams(_nParTot);
+    affectedParams.setZero();
+
+    unsigned nOutliers = 0;  // returned to the caller
+    // start from the strongest outliers.
+    for (auto chi2 = chi2List.rbegin(); chi2 != chi2List.rend(); ++chi2) {
+        if (chi2->chi2 < cut) break;  // because the array is sorted.
+        std::vector<unsigned> indices;
+        indices.reserve(100);  // just there to limit reallocations.
```

Comment:
Usually `std::vector` is better at limiting reallocations than the programmer, especially when magic numbers are involved.